src/org/sosy_lab/java_smt/solvers/z3/Z3UserPropagator.java

```diff
@@ -160,12 +166,16 @@ private long[] extractInfoFromArray(BooleanFormula[] formulaArray) {
     return formulaInfos;
   }
 
-  private BooleanFormula encapsulate(final long z3Expr) {
+  private BooleanFormula encapsulateBoolean(final long z3Expr) {
+    return (BooleanFormula) encapsulate(z3Expr);
+  }
+
+  private Formula encapsulate(final long z3Expr) {
     // Due to pointer alignment, the lowest 2-3 bits are always 0 which can lead to
     // more collisions in the hashmap. To counteract, we fill the lowest bits by rotating the
     // value. The rotation guarantees a bijective transformation.
     return canonicalizer.computeIfAbsent(
-        Long.rotateRight(z3Expr, 3), key -> creator.encapsulateBoolean(z3Expr));
+        Long.rotateRight(z3Expr, 3), key -> creator.encapsulateWithTypeOf(z3Expr));
```

Comment:
We should start a test-run for this, whether more instances can be triggered by extensive usage.
Good spot!